src/core/memory.d

```diff
@@ -810,6 +810,91 @@ struct GC
     }
 }
 
+/**
+ * Pure variants of C's memory allocation functions `malloc`, `calloc`, and
+ * `realloc` that achieve purity by aborting the program on failure so
+ * they never visibly change errno.
+ *
+ * The functions may terminate the program using `onOutOfMemoryError` or
+ * `assert(0)`. These functions' purity guarantees no longer hold if
+ * the program continues execution after catching AssertError or
+ * OutOfMemoryError.
+ *
+ * A similar fail-fast variant of C's memory deallocation function `free`
+ * is not provided due to the lack of a portable and efficient way to test
+ * for failure. If purity is required `pureFree` may be used.
+ *
+ * See_Also:
+ *     $(LREF pureMalloc), $(LREF pureCalloc), $(LREF pureRealloc),
+ *     and $(LREF pureFree) which achieve purity by saving and restoring
+ *     errno.
+ *     $(LINK2 https://dlang.org/library/object/error.html, Error class documentation)
+ *     on the danger of continued execution after catching an Error.
+ */
+void* assertMalloc()(size_t size) @nogc nothrow pure @trusted
+{
+    void* ret = fakePureMalloc(size);
+    // Do not replace `if(condition) assert(0)` with `assert(!condition)`.
+    // The compiler would be permitted to omit the latter in release builds.
+    if (!ret && size)
+        assert(0, "Memory allocation failed");
+    return ret;
+}
+/// ditto
+void* assertCalloc()(size_t nmemb, size_t size) @nogc nothrow pure @trusted
+{
+    void* ret = fakePureCalloc(nmemb, size);
+    // Do not replace `if(condition) assert(0)` with `assert(!condition)`.
+    // The compiler would be permitted to omit the latter in release builds.
+    if (!ret && nmemb && size)
+        assert(0, "Memory allocation failed");
+    return ret;
+}
+/// ditto
+void* assertRealloc()(void* ptr, size_t size) @nogc nothrow pure @system
+{
+    void* ret = fakePureRealloc(ptr, size);
+    // Do not replace `if(condition) assert(0)` with `assert(!condition)`.
+    // The compiler would be permitted to omit the latter in release builds.
+    if (!ret && size)
+        assert(0, "Memory allocation failed");
+    return ret;
+}
+
+///
+@nogc nothrow pure @system unittest
+{
+    void[] allocate(size_t n) pure
+    {
+        return assertMalloc(n)[0 .. n];
+    }
+
+    auto buf = allocate(100);
+    assert(buf.length == 100);
+    pureFree(buf.ptr);
+}
+
+@nogc nothrow pure @system unittest
+{
+    const int errno = fakePureErrno();
```

Comment:
`fakePureErrno` is mangled to `fakePureErrnoImpl` which is not a template.  Therefore this won't work until `fakePureErrnoImpl` is made into a template, or some other solution is devised.

Reply:
Should be fixed now.